File: sdks/go/event.go
package condux

import (
	"net/http"
	"reflect"
)

// Level is the event severity, matching the levels the relay understands. The wire value is the lowercase
// string.
type Level string

const (
	LevelDebug   Level = "debug"
	LevelInfo    Level = "info"
	LevelWarning Level = "warning"
	LevelError   Level = "error"
	LevelFatal   Level = "fatal"
)

// SendResult is the outcome of a delivery attempt sequence. Reporting never panics — inspect OK. Status is
// the last HTTP status seen (0 if every attempt was a network error); Error is the last transport error.
type SendResult struct {
	OK       bool
	Attempts int
	Status   int
	Error    string
}

// The Sentry store wire shape. Serialized to snake_case JSON so the relay parses it like a Sentry SDK.
// Every enrichment field is omitempty, so an unenriched event carries none of those keys at all.
type event struct {
	EventID     string                    `json:"event_id"`
	Timestamp   float64                   `json:"timestamp"`
	Platform    string                    `json:"platform"`
	Level       Level                     `json:"level"`
	Environment string                    `json:"environment,omitempty"`
	Release     string                    `json:"release,omitempty"`
	Message     string                    `json:"message,omitempty"`
	Exception   *exceptions               `json:"exception,omitempty"`
	User        *User                     `json:"user,omitempty"`
	Tags        map[string]string         `json:"tags,omitempty"`
	Contexts    map[string]map[string]any `json:"contexts,omitempty"`
	Breadcrumbs *breadcrumbs              `json:"breadcrumbs,omitempty"`
	Request     *Request                  `json:"request,omitempty"`
	// Modules is the runtime dependency inventory (ADR-0041): the module versions built into this
	// binary. omitempty, so an event that carries none keeps its exact previous wire shape.
	Modules map[string]string `json:"modules,omitempty"`
}

// Request is the request an event happened during. The JSON names are the Sentry store shape the relay
// parses, so query_string is snake_case on purpose.
//
// Deliberately no headers: they carry Cookie and Authorization, and while the relay scrubs sensitive
// keys at ingest, not sending credentials at all is the stronger guarantee.
type Request struct {
	// URL is the path or absolute URL, without the query string.
	URL         string `json:"url,omitempty"`
	Method      string `json:"method,omitempty"`
	QueryString string `json:"query_string,omitempty"`
}

// RequestFrom describes an inbound *http.Request for reporting. Nil-safe, so a handler that captures
// outside a request does not have to branch.
func RequestFrom(r *http.Request) *Request {
	if r == nil {
		return nil
	}
	request := &Request{Method: r.Method}
	if r.URL != nil {
		request.URL = r.URL.Path
		request.QueryString = r.URL.RawQuery
	}
	return request
}

// CaptureContext carries detail belonging to one event rather than to the process.
//
// This exists because the package scope (SetUser, SetTag) is global: a Go server handles requests on
// many goroutines at once, so request detail set there attaches to whichever event is captured next,
// which may belong to a different request. A wrong URL is worse than none, because it sends whoever is
// debugging to the wrong endpoint.
type CaptureContext struct {
	Request *Request
	// Tags merge over the ambient ones, so a per-event tag cannot silently drop the deployment-wide ones.
	Tags map[string]string
}

// Breadcrumbs ride the Sentry {"values": []} envelope, not a bare array.
type breadcrumbs struct {
	Values []Breadcrumb `json:"values"`
}

type exceptions struct {
	Values []exceptionValue `json:"values"`
}

type exceptionValue struct {
	Type       string      `json:"type"`
	Value      string      `json:"value"`
	Mechanism  *mechanism  `json:"mechanism,omitempty"`
	Stacktrace *stacktrace `json:"stacktrace,omitempty"`
}

// A user-invoked capture is a handled capture (Sentry's default). The relay reads mechanism.handled for
// the unhandled badge.
type mechanism struct {
	Type    string `json:"type"`
	Handled bool   `json:"handled"`
}

type stacktrace struct {
	Frames []frame `json:"frames"`
}

type frame struct {
	Filename string `json:"filename,omitempty"`
	Function string `json:"function,omitempty"`
	Module   string `json:"module,omitempty"`
	Lineno   int    `json:"lineno,omitempty"`
	InApp    bool   `json:"in_app"`
}

func toException(err error, handled bool) *exceptions {
	value := "nil error"
	typ := "error"
	if err != nil {
		value = err.Error()
		typ = reflect.TypeOf(err).String()
	}

	ex := exceptionValue{
		Type:      typ,
		Value:     value,
		Mechanism: &mechanism{Type: "generic", Handled: handled},
	}
	if frames := captureStack(); len(frames) > 0 {
		ex.Stacktrace = &stacktrace{Frames: frames}
	}
	return &exceptions{Values: []exceptionValue{ex}}
}

File: backend/src/Condux.Core/Events/Event.cs
namespace Condux.Core.Events;

/// <summary>Severity of an event, mirroring the common SDK levels.</summary>
public enum Level
{
    Unspecified,
    Debug,
    Info,
    Warning,
    Error,
    Fatal,
}

/// <summary>A single stack frame.</summary>
public sealed record Frame
{
    public string? Filename { get; init; }

    /// <summary>The frame's absolute path or URL (Sentry <c>abs_path</c>). Kept for source-map matching by
    /// path when a frame carries no debug id (ADR-0028); a bundled web SDK usually sends the built file URL
    /// here (e.g. https://app/js/app.min.js).</summary>
    public string? AbsPath { get; init; }
    public string? Function { get; init; }
    public string? Module { get; init; }
    public int Lineno { get; init; }
    public int Colno { get; init; }
    public bool InApp { get; init; }
    public string? ContextLine { get; init; }

    /// <summary>Source lines around <see cref="ContextLine"/> (Sentry pre_context/post_context),
    /// powering the code-path view. Scrubbed like any string; empty when the SDK sends none.</summary>
    public IReadOnlyList<string> ContextBefore { get; init; } = [];
    public IReadOnlyList<string> ContextAfter { get; init; } = [];

    /// <summary>Local variable values at this frame (Sentry vars), stringified. Scrubbed by key and
    /// value; empty when the SDK does not capture locals.</summary>
    public IReadOnlyDictionary<string, string> Vars { get; init; } =
        System.Collections.ObjectModel.ReadOnlyDictionary<string, string>.Empty;

    /// <summary>The package or binary image the frame belongs to (Sentry <c>package</c>): the Dart package
    /// for a Dart frame, the loaded binary for a native one.</summary>
    public string? Package { get; init; }

    /// <summary>The resolved symbol name when the SDK sends one separately from <see cref="Function"/>.</summary>
    public string? Symbol { get; init; }

    /// <summary>Native symbolication inputs, sent by the mobile SDKs for frames with no source-level
    /// information. Kept verbatim and unscrubbed (opaque addresses, like a debug id) because they are
    /// discarded irrecoverably otherwise: an event stored without them can never be symbolicated later,
    /// even once a symbolication path exists. Null for a source-level frame.</summary>
    public string? InstructionAddr { get; init; }
    public string? ImageAddr { get; init; }
    public string? SymbolAddr { get; init; }
}

public sealed record Stacktrace
{
    /// <summary>Ordered oldest (outermost) → newest (crashing) frame.</summary>
    public IReadOnlyList<Frame> Frames { get; init; } = [];
}

/// <summary>One exception in a (possibly chained) error.</summary>
public sealed record ExceptionValue
{
    public string? Type { get; init; }
    public string? Value { get; init; }
    public string? Module { get; init; }
    public Stacktrace? Stacktrace { get; init; }

    /// <summary>Whether the error was caught by the app (from the SDK's mechanism); null = unreported.
    /// False is the "unhandled crash" badge.</summary>
    public bool? Handled { get; init; }
}

/// <summary>The user affected by an event. Privacy-first at rest: the scrubber redacts
/// <see cref="Email"/>, drops <see cref="IpAddress"/> entirely, and derives the pseudonymous
/// <see cref="Event.UserKey"/> for distinct "users affected" counts before either happens.</summary>
public sealed record EventUser
{
    public string? Id { get; init; }
    public string? Username { get; init; }
    public string? Email { get; init; }
    public string? IpAddress { get; init; }
}

/// <summary>The HTTP request that produced the error (url/method/query plus headers, all scrubbed —
/// sensitive header keys are redacted).</summary>
public sealed record RequestInfo
{
    public string? Url { get; init; }
    public string? Method { get; init; }
    public string? QueryString { get; init; }
    public IReadOnlyDictionary<string, string> Headers { get; init; } = new Dictionary<string, string>();
}

/// <summary>A breadcrumb — a trail of what happened before the error.</summary>
public sealed record Breadcrumb
{
    public long TimestampUnixMs { get; init; }
    public string? Type { get; init; }
    public string? Category { get; init; }
    public string? Message { get; init; }
    public Level Level { get; init; }

    /// <summary>Structured crumb payload (an http crumb's url/status, a ui crumb's selector...).</summary>
    public IReadOnlyDictionary<string, string> Data { get; init; } = new Dictionary<string, string>();
}

/// <summary>A debug image (Sentry <c>debug_meta.images[]</c>): links a debug id to the built file it
/// identifies, so a stack frame can be matched to its uploaded source map for symbolication (ADR-0028).
/// For a JS source map the <see cref="Type"/> is "sourcemap" and <see cref="CodeFile"/> is the built
/// file's URL.</summary>
public sealed record DebugImage
{
    public string? Type { get; init; }
    public string? CodeFile { get; init; }
    public string? DebugId { get; init; }
}

/// <summary>The normalized error event the relay produces and the consumer stores.</summary>
public sealed record Event
{
    public string EventId { get; init; } = "";
    public long TimestampUnixMs { get; init; }
    public string? Platform { get; init; }
    public Level Level { get; init; }
    public string? Logger { get; init; }
    public string? ServerName { get; init; }
    public string? Release { get; init; }
    public string? Environment { get; init; }
    public string? Transaction { get; init; }
    public string? Message { get; init; }
    public IReadOnlyList<ExceptionValue> Exceptions { get; init; } = [];
    public IReadOnlyList<Breadcrumb> Breadcrumbs { get; init; } = [];
    public IReadOnlyDictionary<string, string> Tags { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Extra { get; init; } = new Dictionary<string, string>();
    public IReadOnlyList<string> Fingerprint { get; init; } = [];
    public string? SdkName { get; init; }
    public string? SdkVersion { get; init; }

    /// <summary>The affected user as captured (scrubbed before storage — see <see cref="EventUser"/>).</summary>
    public EventUser? User { get; init; }

    /// <summary>Stable pseudonymous key for distinct "users affected" counts: a hash of the strongest
    /// user identifier, derived by the scrubber at ingest. Empty when the event carried no user.</summary>
    public string UserKey { get; init; } = "";

    /// <summary>The HTTP request that produced the error, when the SDK sent one.</summary>
    public RequestInfo? Request { get; init; }

    /// <summary>Flattened runtime contexts ("browser" → "Chrome 126", "os", "device", "runtime") — the
    /// facet-friendly form of Sentry's contexts object.</summary>
    public IReadOnlyDictionary<string, string> Contexts { get; init; } = new Dictionary<string, string>();

    /// <summary>Release sub-distribution (Sentry <c>dist</c>).</summary>
    public string? Dist { get; init; }

    /// <summary>Installed dependency versions (Sentry <c>modules</c>) — "which library version broke".</summary>
    public IReadOnlyDictionary<string, string> Modules { get; init; } = new Dictionary<string, string>();

    /// <summary>Debug images (Sentry <c>debug_meta.images[]</c>) mapping a debug id to its built file, for
    /// source-map symbolication (ADR-0028). Empty when the SDK sent none.</summary>
    public IReadOnlyList<DebugImage> DebugImages { get; init; } = [];

    /// <summary>The distributed trace this event belongs to (contexts.trace.trace_id), for APM linking.</summary>
    public string? TraceId { get; init; }
}
